app/src/main/java/io/apicurio/registry/mt/limits/RegistryStorageLimitsEnforcer.java

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2021 Red Hat
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.apicurio.registry.mt.limits;
+
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import io.apicurio.registry.content.ContentHandle;
+import io.apicurio.registry.storage.ArtifactAlreadyExistsException;
+import io.apicurio.registry.storage.ArtifactNotFoundException;
+import io.apicurio.registry.storage.RegistryStorageException;
+import io.apicurio.registry.storage.VersionNotFoundException;
+import io.apicurio.registry.storage.decorator.RegistryStorageDecorator;
+import io.apicurio.registry.storage.dto.ArtifactMetaDataDto;
+import io.apicurio.registry.storage.dto.EditableArtifactMetaDataDto;
+import io.apicurio.registry.types.ArtifactType;
+
+/**
+ * @author Fabian Martinez
+ */
+@ApplicationScoped
+public class RegistryStorageLimitsEnforcer extends RegistryStorageDecorator {
+
+    @Inject
+    TenantLimitsService limitsService;
+
+    @Inject
+    TenantLimitsConfigurationService limitsConfiguration;
+
+    /**
+     * @see io.apicurio.registry.storage.decorator.RegistryStorageDecorator#isConfigured()
+     */
+    @Override
+    public boolean isConfigured() {
+        return limitsConfiguration.isConfigured();
+    }
+
+    /**
+     * @see io.apicurio.registry.storage.decorator.RegistryStorageDecorator#order()
+     */
+    @Override
+    public int order() {
+        return 0;
+    }
+
+    /**
+     * @see io.apicurio.registry.storage.decorator.RegistryStorageDecorator#createArtifact(java.lang.String, java.lang.String, java.lang.String, io.apicurio.registry.types.ArtifactType, io.apicurio.registry.content.ContentHandle)
+     */
+    @Override
+    public CompletionStage<ArtifactMetaDataDto> createArtifact(String groupId, String artifactId,
+            String version, ArtifactType artifactType, ContentHandle content)
+            throws ArtifactAlreadyExistsException, RegistryStorageException {
+
+        LimitsCheckResult r = limitsService.canCreateArtifact(null);
```

Comment:
Maybe here you can use a functional approach for the validations since you're repeating essentially the same code across the enforcer (check, execute, get result).

Reply:
Functional approach is a nice idea.  +1 to that